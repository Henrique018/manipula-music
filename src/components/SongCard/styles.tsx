import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
	${({ theme }) => css`
		background-color: ${theme.colors.tertiary};
		border-radius: ${theme.border.radius};
		padding: ${theme.spacings.xsmall};
		position: relative;
		cursor: pointer;
		transition: background-color 400ms ease-in-out;

		&:hover {
			background-color: ${theme.colors.secondary};
		}
	`}
`;

export const Image = styled.img`
	width: 100%;
	border-radius: 0.3rem;
`;

export const PlayBtn = styled.button`
	${({ theme }) => css`
		position: absolute;
		top: 19rem;
		left: 3rem;

		display: flex;
		align-items: center;
		justify-content: center;
		color: ${theme.colors.white};
		border-radius: 50%;
		background: ${theme.colors.primary};
		border: 0;
	`}
`;

export const Content = styled.div``;

export const Title = styled.h2`
	${({ theme }) => css`
		font-size: ${theme.font.sizes.medium};
	`}
`;
export const Info = styled.div`
	display: flex;
	justify-content: space-between;
	align-items: center;
`;

export const ArtistName = styled.p`
	${({ theme }) => css`
		font-size: ${theme.font.sizes.small};
		display: inline;
		margin-right: ${theme.spacings.xxsmall};
	`}
`;

export const Duration = styled.span`
	${({ theme }) => css`
		font-size: ${theme.font.sizes.xsmall};
		font-weight: 300;
	`}
`;
export const Buttons = styled.div`
	display: flex;
	align-items: center;
`;

export const DeezerBtn = styled.a``;

export const FavoriteBtn = styled.button`
	background: transparent;
	margin-left: 1rem;
	border: 0;
`;
