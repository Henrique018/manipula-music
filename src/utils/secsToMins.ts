function secsToMins(seconds: number): string {
	// 1 sec == 0.016 min
	const mins = (seconds * 0.016).toFixed(2);

	return mins.replace('.', ':');
}

export default secsToMins;
