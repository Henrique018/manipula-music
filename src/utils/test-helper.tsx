import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';

import theme from 'styles/theme';

export const renderWithProviders = (
	ui: ReactElement,
	route = '/'
): RenderResult => {
	window.history.pushState({}, 'Test page', route);
	return render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, {
		wrapper: BrowserRouter
	});
};
