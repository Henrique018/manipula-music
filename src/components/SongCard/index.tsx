import { useDispatch } from 'react-redux';
import { LinkExternal, Star } from '@styled-icons/boxicons-regular';
import { Star as FilledStar } from '@styled-icons/boxicons-solid';

import { AppDispatch } from 'redux/store';
import { addToFav, removeFromFav } from 'redux/songsSlicer';
import * as S from './styles';
import AudioButton from 'components/AudioButton';

export type SongCardProps = {
	id: number;
	title: string;
	artistName: string;
	duration: string;
	albumCover: string;
	link: string;
	favorite?: boolean;
	preview: string;
};

const SongCard = ({
	id,
	title,
	artistName,
	duration,
	albumCover,
	link,
	favorite = false,
	preview
}: SongCardProps) => {
	const dispatch = useDispatch<AppDispatch>();

	const onFav = (id: number) => {
		if (favorite) {
			dispatch(removeFromFav({ id }));
		} else {
			dispatch(
				addToFav({
					id,
					title,
					artistName,
					albumCover,
					duration,
					link,
					favorite: true,
					preview
				})
			);
		}
	};

	return (
		<S.Wrapper>
			<S.Image src={albumCover} alt={`${artistName}'s album cover`} />
			<AudioButton audioSrc={preview} />
			<S.Content>
				<S.Title>{title}</S.Title>
				<S.Info>
					<div>
						<S.ArtistName>{artistName}</S.ArtistName>
						<S.Duration>{duration}</S.Duration>
					</div>
					<S.Buttons>
						<S.DeezerBtn
							href={link}
							target="_blank"
							rel="noopener noreferrer"
							title="Listen on Deezer"
						>
							<LinkExternal size={31} color="#E94560" />
						</S.DeezerBtn>

						<S.FavoriteBtn onClick={() => onFav(id)}>
							{favorite ? (
								<FilledStar
									size={32}
									color="#E94560"
									aria-label="Remove from favorites"
									title="Remove from favorites"
								/>
							) : (
								<Star
									size={32}
									color="#E94560"
									aria-label="Add to favorites"
									title="Add to favorites"
								/>
							)}
						</S.FavoriteBtn>
					</S.Buttons>
				</S.Info>
			</S.Content>
		</S.Wrapper>
	);
};

export default SongCard;
