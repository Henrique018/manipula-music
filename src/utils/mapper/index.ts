import secsToMins from 'utils/secsToMins';

type ApiDataProps = {
	data: {
		id: number;
		title: string;
		duration: number;
		link: string;
		preview: string;
		artist: {
			name: string;
		};
		album: {
			cover_medium: string;
		};
	}[];
};

const mapToSongCard = (apiData: Partial<ApiDataProps>) => {
	return apiData.data?.map((track) => ({
		id: track.id,
		title: track.title,
		artistName: track.artist.name,
		duration: secsToMins(track.duration),
		albumCover: track.album.cover_medium,
		link: track.link
	}));
};

export default mapToSongCard;
