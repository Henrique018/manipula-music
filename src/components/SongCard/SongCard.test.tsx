import { render, screen } from 'utils/test-helper';
import SongCard, { SongCardProps } from '.';

const mock = {
	id: 1,
	title: 'Feel Good Inc.',
	artistName: 'Gorillaz',
	duration: '3:00',
	albumCover:
		'https://cdns-images.dzcdn.net/images/cover/8dd837db7a1c6acf7acb20e26c0452c3/250x250-000000-80-0-0.jpg',
	link: 'https://www.deezer.com/track/3129407',
	preview: 'songpreview',
	favorite: false
} as SongCardProps;

describe('<SongCard>', () => {
	it('should have img, title, artist, play, external link and favorite', () => {
		const { container } = render(<SongCard {...mock} />);

		expect(screen.getByAltText(/gorillaz's album cover/i)).toBeInTheDocument();

		expect(
			screen.getByRole('heading', { name: /Feel Good Inc/i })
		).toBeInTheDocument();

		expect(screen.getByText(/Gorillaz/i)).toBeInTheDocument();
		expect(screen.getByText(/3:00/i)).toBeInTheDocument();

		expect(
			screen.getByRole('button', { name: /listen song preview/i })
		).toBeInTheDocument();

		expect(
			screen.getByRole('link', { name: /listen on deezer/i })
		).toBeInTheDocument();

		expect(screen.getByLabelText(/add to favorites/i)).toBeInTheDocument();

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render a filled star if its a favorite song', () => {
		render(<SongCard {...mock} favorite />);

		expect(screen.getByLabelText(/remove from favorites/i)).toBeInTheDocument();
	});
});
