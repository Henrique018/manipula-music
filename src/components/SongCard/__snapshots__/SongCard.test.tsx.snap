// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongCard> should have img, title, artist, play, external link and favorite 1`] = `
<div
  class="sc-bdnxRM bjyjWg"
>
  <img
    alt="Gorillaz's album cover"
    class="sc-gtsrHT gEVLss"
    src="https://cdns-images.dzcdn.net/images/cover/8dd837db7a1c6acf7acb20e26c0452c3/250x250-000000-80-0-0.jpg"
  />
  <button
    class="sc-dlnjwi hNmXCY"
    title="Listen the preview"
  >
    <svg
      aria-hidden="true"
      class="StyledIconBase-ea9ulj-0 fBxgqm"
      color="#fff"
      fill="currentColor"
      focusable="false"
      height="36"
      viewBox="0 0 24 24"
      width="36"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 6v12l10-6z"
      />
    </svg>
  </button>
  <div
    class="sc-hKFxyN"
  >
    <h2
      class="sc-eCApnc eLhFMg"
    >
      Feel Good Inc.
    </h2>
    <div
      class="sc-jSFjdj gNdbja"
    >
      <div>
        <p
          class="sc-gKAaRy brovIv"
        >
          Gorillaz
        </p>
        <span
          class="sc-iCoGMd jWILFb"
        >
          300
        </span>
      </div>
      <div
        class="sc-fujyAs VGLZA"
      >
        <a
          class="sc-pNWdM"
          href="https://www.deezer.com/track/3129407"
          rel="noopener noreferrer"
          target="_blank"
          title="Listen on Deezer"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 fBxgqm"
            color="#E94560"
            fill="currentColor"
            focusable="false"
            height="32"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m13 3 3.293 3.293-7 7 1.414 1.414 7-7L21 11V3z"
            />
            <path
              d="M19 19H5V5h7l-2-2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2h14c1.103 0 2-.897 2-2v-5l-2-2v7z"
            />
          </svg>
        </a>
        <button
          class="sc-jrsJWt fLZph"
        >
          <svg
            aria-label="Add to favorites"
            class="StyledIconBase-ea9ulj-0 fBxgqm"
            color="#E94560"
            fill="currentColor"
            focusable="false"
            height="32"
            role="img"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Add to favorites
            </title>
            <path
              d="m6.516 14.323-1.49 6.452a.998.998 0 0 0 1.529 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082a1 1 0 0 0-.59-1.74l-5.701-.454-2.467-5.461a.998.998 0 0 0-1.822 0L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.214 4.107zm2.853-4.326a.998.998 0 0 0 .832-.586L12 5.43l1.799 3.981a.998.998 0 0 0 .832.586l3.972.315-3.271 2.944c-.284.256-.397.65-.293 1.018l1.253 4.385-3.736-2.491a.995.995 0 0 0-1.109 0l-3.904 2.603 1.05-4.546a1 1 0 0 0-.276-.94l-3.038-2.962 4.09-.326z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
